### PR TITLE
[SW-217] Add unit testing framework to RSparkling

### PR DESCRIPTION
**Pull request by navdeep-G**
*This PR adds a unit testing framework for RSparkling (Note: It is not complete, but a good start)
*To run simply call `./gradlew testR` in top directory of sparkling water

### CODE REVIEW

**Comment by navdeep-G** [comment]
Failed checks are probably due to R dependencies not in travis?

**Comment by jakubhava** [comment]
```
* What went wrong:
Execution failed for task ':sparkling-water-r:testR'.
> Process 'command 'python'' finished with non-zero exit value 1

```

Seems like python not available

**Comment by mmalohlava** [comment]
Probably this one: `OSError: [Errno 2] No such file or directory`

**Comment by navdeep-G** [comment]
Weird, ran fine locally. Maybe its a path issue?

**Comment by navdeep-G** [comment]
Can one of you try it locally? Would be good to know of any issues.

**Comment by navdeep-G** [comment]
Any updates?

**Comment by mdymczyk** [comment]
@navdeep-G I'm also getting errors locally:

```
:sparkling-water-r:testR
dyld: Library not loaded: /usr/local/opt/readline/lib/libreadline.6.dylib
  Referenced from: /usr/local/Cellar/r/3.2.4_1/R.framework/Versions/3.2/Resources/lib/libR.dylib
  Reason: image not found
Traceback (most recent call last):
Running the following test: test_as_h2o_frame_df.R
  File "run.py", line 18, in <module>
    run_all()
  File "run.py", line 14, in run_all
    subprocess.check_call(['R','-f', file], shell=False)
  File "/Users/mateusz/Dev/tools/anaconda/lib/python2.7/subprocess.py", line 541, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['R', '-f', 'test_as_h2o_frame_df.R']' returned non-zero exit status -5
:sparkling-water-r:testR FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':sparkling-water-r:testR'.
> Process 'command 'python'' finished with non-zero exit value 1
```

Running ```R -f test_as_h2o_frame_df.R``` results in:

```
dyld: Library not loaded: /usr/local/opt/readline/lib/libreadline.6.dylib
  Referenced from: /usr/local/Cellar/r/3.2.4_1/R.framework/Versions/3.2/Resources/lib/libR.dylib
  Reason: image not found
```

Any ideas?

**Comment by jakubhava** [comment]
I got the same old one 

```> #Create a spark connection
> sc <- spark_connect(master = "local")
Error in force(code) :
  Failed while connecting to sparklyr to port (8880) for sessionid (1637): Gateway in port (8880) did not respond.
    Path: /Users/kuba/programs/spark-2.0.0-bin-hadoop2.7/bin/spark-submit
    Parameters: --class, sparklyr.Backend, --packages, 'com.databricks:spark-csv_2.11:1.3.0','ai.h2o:sparkling-water-core_2.11:1.6.7','ai.h2o:sparkling-water-ml_2.11:1.6.7','ai.h2o:sparkling-water-repl_2.11:1.6.7', '/usr/local/lib/R/3.2/site-library/sparklyr/java/sparklyr-2.0-2.11.jar', 8880, 1637
    Traceback:
      shell_connection(master = master, spark_home = spark_home, app_name = app_name, version = version, hadoop_version = hadoop_version, shell_args = shell_args, config = config, service = FALSE, extensions = extensions)
      start_shell(master = master, spark_home = spark_home, spark_version = version, app_name = app_name, config = config, jars = spark_config_value(config, "spark.jars.default", list()), packages = spark_config_value(config, "sparklyr.defaultPackages"), extensions = extensions, environment = environment, shell_args = shell_args, service = service)
      tryCatch({
    gatewayInfo <- spark_connect_gateway(gatewayAddress, gatewayPort, sessionId, config = config, isStarting = TRUE)
}, error = function(e) {
    abort_shell(paste("Failed while connecting to sparklyr to port (", gatewayPort, ") for sessionid (", sessionId, "): ", e$message, sep = ""), spark_submit_path, shell_args, output_file, error_file)
})
```

**Comment by jakubhava** [comment]
It's related to this one - https://github.com/h2oai/sparkling-water/pull/116#pullrequestreview-6977793. 

In each test we would need to specify sparkling-water version. So it would make sense to specify the version in `r/build.gradle` as environmental property and then in tests fetch it from the environment like 
```options(rsparkling.sparklingwater.version = GET_ENV(SPARKLING_WATER_VERSION)```

This would just ensure that the version is configured at one place and we don't have to update version in all tests when new version of sparkling-water comes.

**Comment by jakubhava** [comment]
For me the tests are running correctly when doing `./gradlew build -x check testR`

Though one test is failing:
```
> #Run test
> test_h2o_name()
Error in as_h2o_frame(sc, mtcars_tbl, name = "frame1") :
  unused argument (name = "frame1")
Calls: test_h2o_name -> as_h2o_frame
Execution halted
Traceback (most recent call last):
  File "run.py", line 18, in <module>
    run_all()
Running the following test: test_as_h2o_frame_df.R
  File "run.py", line 14, in run_all
    subprocess.check_call(['R','-f', file], shell=False)
Running the following test: test_df_bool_h2o_frame.R
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
```

**Comment by jakubhava** [comment]
@mdymczyk it seems line environmental issue - see http://stackoverflow.com/questions/22631026/dyld-library-not-loaded-usr-local-lib-libpng16-16-dylib-with-anything-php-rel/22631734#22631734

**Comment by mdymczyk** [comment]
Right updating my brew helped but seems my R env is a mess so will take a bit of time ;-)

In the meantime it seems travis build passed so it should be ok to merge?

**Comment by jakubhava** [comment]
@mdymczyk Travis build is skipping R tests

I wrote 2 comments above - 1 tests is still failing and need to be fixed

Also for the future, sparkling-water version should be probably set in r/build.gradle. I know that it doesn't apply to this particular change, since this version is still using hardcoded sparkling-water version. I just wrote it here because it seems to be relevant to the tests. We can create a JIRA for it and implement it once this and https://github.com/h2oai/sparkling-water/pull/116 is merged.

**Comment by jangorecki** [comment]
`tests/run.py` could goes outside of R package. Package check already runs every *.R in tests, if we want to it run them from gradlew then it should be ok to have run.py outside of rsparkling. Didn't read the discussion yet, will take a loot later today.

File: r/rsparkling/tests/test_as_h2o_frame_df.R [hunk 1]
```diff
@@ -0,0 +1,23 @@
+library(testthat)
+library(rsparkling)
+library(dplyr)
```
Comment (jangorecki) [inline r/rsparkling/tests/test_as_h2o_frame_df.R]: Here and in the other tests we could also load `sparklyr` as we use `spark_connect()` from `sparklyr`. Now it rely on that fact that `rsparkling depends on sparklyr, avoiding depends in favour of imports is a good practice and prevents various issues for users/downstream packages.

File: r/rsparkling/tests/run.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""
```
Comment (jangorecki) [inline r/rsparkling/tests/run.py]: `tests/run.py` is not a part of rsparkling package, it should be placed outside of R package. Package check already runs every *.R in tests directory. If we want to it run them from gradlew then it should be fine to have run.py outside of rsparkling directory.

**Comment by navdeep-G** [comment]
All, this whole thing will get revamped today. Just give me time.

**Pull request closed without merging**